### PR TITLE
Fixed #745 - reduce visibility of classes

### LINKED ISSUE #745: Hide all classes that should be private
Several classes are showing up in Scaladoc that are not directly necessary or exposed. Examples include: `ZIOFunctions`, `ZIO_E_Any`, etc., and probably various classes in the `scalaz.zio.internal` package that will never be exposed because they are only used internally by `FiberContext`.

To present clear and useful Scaladoc, all internal classes / traits should be hidden from the user, so that only types intended to be used by end-users appear in the documentation.

This can be done, at least, by making classes / traits private where relevant.

### PR DESCRIPTION
**Pull request by GodPlaysChess**
Hide classes and traits which either should not be exposed or present in scaladoc.

I made ADTs for ZIO package private, since the they are only used internally, and should not be pattern matched.
Also I made MutableConcurrentQueue package protected, since it is not supposed to be used outside..

### CODE REVIEW

File: core/shared/src/main/scala/scalaz/zio/internal/ExecutionMetrics.scala [hunk 1]
```diff
@@ -16,7 +16,7 @@
 
 package scalaz.zio.internal
 
-trait ExecutionMetrics {
+private[internal] trait ExecutionMetrics {
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/internal/ExecutionMetrics.scala]: This trait isn't exposed anywhere???

File: core/shared/src/main/scala/scalaz/zio/internal/Stack.scala [hunk 2]
```diff
@@ -19,7 +19,7 @@ package scalaz.zio.internal
 /**
  * A very fast, growable/shrinkable, mutable stack.
  */
-final class Stack[A <: AnyRef]() {
+private[zio] final class Stack[A <: AnyRef]() {
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/internal/Stack.scala]: Don't forget `StackBool` while you're at it!

File: core/shared/src/main/scala/scalaz/zio/internal/ExecutionMetrics.scala (hunk 1 above)
Comment (GodPlaysChess) [reply core/shared/src/main/scala/scalaz/zio/internal/ExecutionMetrics.scala]: The only place where you might need to implement it - in `Executor`.
But you're right, I've missed it - will revert.

File: core/shared/src/main/scala/scalaz/zio/internal/Stack.scala (hunk 2 above)
Comment (GodPlaysChess) [reply core/shared/src/main/scala/scalaz/zio/internal/Stack.scala]: Right, I've thought that private constructor is sufficient :(

**Comment by GodPlaysChess** [comment]
There are few such things, which I was not sure about, such as
`<*>` object, which is just a sugar to deconstruct tuples.
or `Live` objects - which, I think makes sense to expose, but they lack documentation.
Besides that I did not spot anything.

Also, I have troubles to fix Lint build. I formatted `ZIO` with scalafmt, but the build still fails without providing any specific problem, besides the file is not formatted properly :/. 

Should I look more into it? Mb there's a way to set it up, so it shows the exact problem.

File: core/shared/src/main/scala/scalaz/zio/internal/Stack.scala (hunk 2 above)
Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/internal/Stack.scala]: If the class itself is not private, it will appear in the generated docs... 😦

**Comment by jdegoes** [comment]
@GodPlaysChess Try running `fmt` from SBT. It is an alias command that will format the build file too, which is probably what was missed.

**Comment by GodPlaysChess** [comment]
Thx, it worked!
Dotty failures seem to be irrelevant to the changes I've made. 
Should I look into them or it is known issue?

**Comment by jdegoes** [comment]
The Dotty error:

```
[error] -- [E008] Member Not Found Error: /home/travis/build/scalaz/scalaz-zio/core/shared/src/main/scala/scalaz/zio/ZManaged.scala:41:12 
[error] 41 |    reserve.bracket(_.release)(_.acquire.flatMap(f))
[error]    |    ^^^^^^^^^^^^^^^
[error]    |value bracket is not a member of scalaz.zio.ZIO[R, E, scalaz.zio.Reservation[R, E, A]] - did you mean reserve.bracket_?
[error] -- [E081] Syntax Error: /home/travis/build/scalaz/scalaz-zio/core/shared/src/main/scala/scalaz/zio/ZManaged.scala:41:20 
[error] 41 |    reserve.bracket(_.release)(_.acquire.flatMap(f))
[error]    |                    ^
[error]    |Missing parameter type
[error]    |
[error]    |The argument types of an anonymous function must be fully known. (SLS 8.5)
[error]    |Expected type: ?
[error]    |Missing type for parameter _$207
[error] -- [E081] Syntax Error: /home/travis/build/scalaz/scalaz-zio/core/shared/src/main/scala/scalaz/zio/ZManaged.scala:41:31 
[error] 41 |    reserve.bracket(_.release)(_.acquire.flatMap(f))
[error]    |                               ^
[error]    |Missing parameter type
[error]    |
[error]    |The argument types of an anonymous function must be fully known. (SLS 8.5)
[error]    |Expected type: ?
[error]    |Missing type for parameter _$208
[warn] -- Warning: /home/travis/build/scalaz/scalaz-zio/core/shared/src/main/scala/scalaz/zio/stm/STM.scala:346:6 
[warn] 346 |      final case object Retry               extends TRez[Nothing, Nothing]
[warn]     |      ^^^^^
[warn]     |      final modifier is redundant for objects
[warn] four warnings found
[error] 6 errors found
[error] (coreJVM / Compile / compileIncremental) Compilation failed
[error] Total time: 20 s, completed Apr 16, 2019 2:07:49 PM
```

This means that things are not visible which should be visible. Probably one (all?) of the helper `Bracket` classes in `ZIO` object used for curried / type-inferred syntax.

**Comment by jdegoes** [comment]
@GodPlaysChess Thank you for your first contribution (of many, I hope!), and congratulations on becoming a ZIO contributor! 😄 

giphy

**Pull request merged**